x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/execution_plan/required_permissions.ts

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ActionsByType } from './types';
+
+export interface RequiredPermissions {
+  cluster: string[];
+  index: Record<string, string[]>;
+}
+
+export function mergeRequiredPermissions(permissions: RequiredPermissions[]): RequiredPermissions {
+  const clusterSet = new Set<string>();
+  const indexPermissions: Record<string, Set<string>> = {};
+
+  permissions.forEach((permission) => {
+    // Add cluster permissions to the set
+    permission.cluster.forEach((perm) => clusterSet.add(perm));
+
+    // Process index permissions
+    Object.entries(permission.index).forEach(([index, perms]) => {
+      if (!indexPermissions[index]) {
+        indexPermissions[index] = new Set<string>();
+      }
+      perms.forEach((perm) => indexPermissions[index].add(perm));
+    });
+  });
+
+  // Convert sets back to arrays for the return value
+  const result: RequiredPermissions = {
+    cluster: Array.from(clusterSet),
+    index: {},
+  };
+
+  Object.entries(indexPermissions).forEach(([index, permSet]) => {
+    result.index[index] = Array.from(permSet);
+  });
+
+  return result;
+}
+
+export function getRequiredPermissionsForActions(actions: ActionsByType): RequiredPermissions {
+  const permissions: RequiredPermissions[] = [];
+
+  if (actions.upsert_component_template.length > 0) {
```

Comment:
This a little type headache, but could we do the same unpacking check we do for `executeActions`? 

Just to make sure that we actually check **all** action types, since else if we add a new type there won't be type feedback about updating this code.